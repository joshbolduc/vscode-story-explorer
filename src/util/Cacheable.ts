/**
 * An object that lazily performs a predefined operation and caches its result.
 */
export class Cacheable<T> {
  private cachedResult?: { result: T } | undefined;

  /**
   * Initialize a new Cacheable with a given callback.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public constructor(private callback: () => T) {}

  /**
   * Get the result of the operation, invoking it if necessary. If the operation
   * has already been invoked, its result is returned.
   *
   * @returns The result of the operation, cached or otherwise.
   */
  public getResult() {
    if (!this.cachedResult) {
      this.cachedResult = { result: this.callback() };
    }

    return this.cachedResult.result;
  }

  /**
   * Get the cached result of the operation, if any, or a default value if the
   * result has not already been determined.
   */
  public getCachedResult(): T | undefined;
  public getCachedResult<U>(defaultValue: U): T | U;
  public getCachedResult<U>(defaultValue?: U) {
    return this.cachedResult ? this.cachedResult.result : defaultValue;
  }

  /**
   * Replace the callback used to computed the cached result. The cached result,
   * if any, is cleared.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public setCallback(callback: () => T) {
    this.clear();
    this.callback = callback;
  }

  /**
   * Clears the cached result. The next time a result is requested, the
   * operation will be reinvoked and its return value will be cached again.
   */
  public clear() {
    this.cachedResult = undefined;
  }
}
