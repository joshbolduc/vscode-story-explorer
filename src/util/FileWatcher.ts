import type { Disposable, FileSystemWatcher, GlobPattern, Uri } from 'vscode';
import { workspace } from 'vscode';

export type WatchListener = (
  uri: Uri,
  type: 'create' | 'change' | 'delete',
) => void;

export class FileWatcher {
  private fsWatcher?: FileSystemWatcher | undefined;

  private readonly glob: GlobPattern;

  private readonly listenerDisposables: Disposable[] = [];

  public constructor(
    patternOrUri: GlobPattern | Uri,
    callback: WatchListener,
    ignoreCreate = false,
    ignoreChange = false,
    ignoreDelete = false,
  ) {
    this.glob =
      typeof patternOrUri === 'string' || 'pattern' in patternOrUri
        ? patternOrUri
        : patternOrUri.fsPath;
    this.init(callback, ignoreCreate, ignoreChange, ignoreDelete);
  }

  public dispose() {
    this.listenerDisposables.forEach((disposable) => {
      disposable.dispose();
    });
    this.fsWatcher?.dispose();
    this.fsWatcher = undefined;
  }

  private init(
    callback: WatchListener,
    ignoreCreate: boolean,
    ignoreChange: boolean,
    ignoreDelete: boolean,
  ) {
    this.fsWatcher = workspace.createFileSystemWatcher(
      this.glob,
      ignoreCreate,
      ignoreChange,
      ignoreDelete,
    );

    this.listenerDisposables.push(
      this.fsWatcher.onDidChange((uri) => {
        callback(uri, 'change');
      }),
      this.fsWatcher.onDidCreate((uri) => {
        callback(uri, 'create');
      }),
      this.fsWatcher.onDidDelete((uri) => {
        callback(uri, 'delete');
      }),
    );
  }
}
