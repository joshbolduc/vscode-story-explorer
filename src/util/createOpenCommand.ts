import type { TextDocumentShowOptions, Uri } from 'vscode';

interface OpenCommandParams {
  title?: string;
  tooltip?: string;
  arguments: readonly [uri: Uri, options?: TextDocumentShowOptions | undefined];
}

export interface OpenCommandStaticProperties {
  command: 'vscode.open';
}

export const createOpenInEditorCommand = (
  uri: Uri,
  options?: TextDocumentShowOptions,
) => {
  return createOpenCommand({
    title: 'Open',
    tooltip: 'Open in Editor',
    arguments: [uri, options],
  });
};

export const createOpenCommand = <T extends OpenCommandParams>({
  title,
  tooltip,
  arguments: args,
}: T) => {
  return {
    command: 'vscode.open',
    title,
    tooltip,
    arguments: args,
  } as T & OpenCommandStaticProperties;
};
