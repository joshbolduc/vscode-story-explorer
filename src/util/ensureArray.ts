export const ensureArray = <T>(valueOrArray: T | T[]): T[] => {
  if (Array.isArray(valueOrArray)) {
    return valueOrArray;
  }

  return [valueOrArray];
};
