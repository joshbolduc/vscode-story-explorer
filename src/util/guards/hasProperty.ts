export const hasProperty =
  <T extends string>(propertyName: T) =>
  (v: unknown): v is Record<T, unknown> =>
    !!(v && typeof v === 'object' && propertyName in v);
