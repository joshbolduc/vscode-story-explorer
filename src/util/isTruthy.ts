type Falsy = false | 0 | -0 | 0n | '' | null | undefined;

export const isTruthy = <T>(t: T | Falsy): t is Exclude<T, Falsy> => !!t;
