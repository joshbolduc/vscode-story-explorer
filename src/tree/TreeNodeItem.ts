import {
  ExtensionContext,
  TreeItem,
  TreeItemCollapsibleState,
  window,
  workspace,
} from 'vscode';
import { createOpenInEditorCommand } from '../util/createOpenCommand';
import { getIconPath } from '../util/getIconPath';
import type { TreeNode } from './TreeNode';
import { findEntryByUri } from './findEntryByUri';

const getDefaultCollapsedState = (
  entry: TreeNode,
): TreeItemCollapsibleState => {
  if (entry.type === 'story' || entry.getChildren().length === 0) {
    return TreeItemCollapsibleState.None;
  }

  const isTopLevel = !entry.parent;
  if (isTopLevel) {
    return TreeItemCollapsibleState.Expanded;
  }

  const activeUri = window.activeTextEditor?.document.uri;
  const hasActiveFile = activeUri && !!findEntryByUri([entry], activeUri);
  if (hasActiveFile) {
    return TreeItemCollapsibleState.Expanded;
  }

  return TreeItemCollapsibleState.Collapsed;
};

export class TreeNodeItem extends TreeItem {
  private constructor(
    node: TreeNode,
    context: ExtensionContext,
    label: string,
  ) {
    super(label, getDefaultCollapsedState(node));

    // If kind has multiple files, arbitrarily pick the first one
    const file = node.type === 'kind' ? node.files[0] : undefined;

    const uri = file?.getUri();

    if (uri) {
      this.id = uri.toString();

      const relativePath = workspace.asRelativePath(uri);
      this.description =
        node.type === 'kind' && node.files.length > 1
          ? `${node.files.length} files Â· ${relativePath}`
          : relativePath;
      this.resourceUri = uri;
      this.command = createOpenInEditorCommand(uri);
    }

    const effectiveStory =
      node.type === 'story' ? node.story : node.getEffectiveStory(false);

    if (effectiveStory) {
      this.id = effectiveStory.id;
      this.command = effectiveStory.getOpenCommand();
      this.contextValue = 'story';
    }

    const iconName = node.getIconName();
    if (iconName) {
      this.iconPath = getIconPath(iconName, context);
    }
  }

  public static fromTreeNode(node: TreeNode, context: ExtensionContext) {
    const label = node.name.trim();
    const item = new TreeNodeItem(node, context, label);

    return item;
  }
}
