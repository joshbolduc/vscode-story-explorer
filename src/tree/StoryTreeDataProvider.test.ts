import { basename, extname } from 'path';
import { describe, expect, it } from 'vitest';
import {
  Command,
  TextDocumentShowOptions,
  TreeItemCollapsibleState,
  Uri,
} from 'vscode';
import { addSerializer } from '../../test/util/addSerializer';
import {
  getTreeRoots,
  TreeItemRepresentation,
} from '../../test/util/getTreeRoots';
import { hasProperty } from '../util/guards/hasProperty';
import { isTruthy } from '../util/guards/isTruthy';
import { TreeNodeItem } from './TreeNodeItem';

const isTreeItemRepresentation = (v: unknown): v is TreeItemRepresentation =>
  hasProperty('isTreeItem')(v) && v.isTreeItem === true;

addSerializer({
  test: (v): v is TreeNodeItem => v instanceof TreeNodeItem,
  serialize: (v, config, indentation, depth, refs, printer) => {
    const stringifyCommand = (command: Command) => {
      if (
        command.command === 'vscode.open' &&
        command.title === 'Open' &&
        command.tooltip === 'Open in Editor'
      ) {
        const [file, options = {}] = command.arguments as
          | [Uri]
          | [Uri, TextDocumentShowOptions | undefined];

        return [
          'Open file',
          file !== v.resourceUri && file.toString(),
          options.selection &&
            printer(options.selection, config, indentation, depth, refs),
        ]
          .filter(isTruthy)
          .join(' ');
      }

      return printer(command, config, `${indentation}  `, depth, refs);
    };

    return [
      `- ${v.label!.toString()}`,
      v.id && `Id: ${v.id}`,
      v.resourceUri && `File: ${v.resourceUri.toString()}`,
      v.contextValue && `Context: ${v.contextValue}`,
      v.collapsibleState !== undefined &&
        v.collapsibleState !== TreeItemCollapsibleState.None &&
        TreeItemCollapsibleState[v.collapsibleState],
      v.iconPath &&
        `Icon: ${basename(v.iconPath.toString()).slice(
          0,
          -extname(v.iconPath.toString()).length,
        )}`,
      v.command && `Command: ${stringifyCommand(v.command)}`,
    ]
      .filter(isTruthy)
      .join(`\n${indentation}  `);
  },
});

addSerializer({
  test: (v): v is TreeItemRepresentation[] =>
    Array.isArray(v) && v.every(isTreeItemRepresentation),
  serialize: (v, config, indentation, depth, refs, printer) => {
    return v
      .map((item) => {
        return printer(item, config, indentation, depth, refs);
      })
      .join(`\n\n${indentation}`);
  },
});

addSerializer({
  test: isTreeItemRepresentation,
  serialize: (
    { item, children },
    config,
    indentation,
    depth,
    refs,
    printer,
  ) => {
    const childrenStr =
      !children || children.length === 0
        ? ''
        : `\n\n${indentation}  ${printer(
            children,
            config,
            `${indentation}  `,
            depth,
            refs,
          )}`;

    return `${printer(item, config, indentation, depth, refs)}${childrenStr}`;
  },
});

describe('StoryTreeDataProvider', () => {
  [
    { version: '6' as const, configDir: '.config-unittest' },
    { version: '7' as const, configDir: '.storybook' },
  ].forEach(({ version, configDir }) =>
    it(`creates tree from story files for v${version}`, async () => {
      const rootChildren = await getTreeRoots(version, configDir);
      expect(rootChildren).toMatchSnapshot();
    }),
  );
});
