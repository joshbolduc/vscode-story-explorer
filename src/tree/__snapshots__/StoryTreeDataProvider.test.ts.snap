// Vitest Snapshot v1

exports[`StoryTreeDataProvider > creates tree from story files 1`] = `
- MDX Docs at top level
  Id: mdx-docs-at-top-level--page
  File: file:///mock/basedir/project/v6/src/stories/DocsOnlyTop.stories.mdx
  Context: story
  Icon: document
  Command: Open file [Ln 2, Col 0]

- MDX with stories, top
  Id: mdx-with-stories-top--mdx-with-stories-top
  File: file:///mock/basedir/project/v6/src/stories/MdxWithStoriesTop.stories.mdx
  Context: story
  Icon: bookmark
  Command: Open file [Ln 5, Col 0]

- Top
  Id: top
  File: file:///mock/basedir/project/v6/src/stories/Top.stories.js
  Context: story
  Expanded
  Icon: component
  Command: Open file [Ln 4, Col 15]

  - Top Level Story
    Id: top--top-level-story
    Context: story
    Icon: bookmark
    Command: Open file file:///mock/basedir/project/v6/src/stories/Top.stories.js [Ln 11, Col 13]

- Top Hoisted
  Id: top-hoisted--top-hoisted
  File: file:///mock/basedir/project/v6/src/stories/TopHoisted.stories.js
  Context: story
  Icon: bookmark
  Command: Open file [Ln 11, Col 13]

- Custom title prefix
  Expanded

  - Second level
    Collapsed
    Icon: folder

    - csf
      Collapsed
      Icon: folder

      - EmptyDefaultExport
        Id: custom-title-prefix-second-level-csf-emptydefaultexport
        File: file:///mock/basedir/project/v6/src/autoTitle/csf/EmptyDefaultExport.stories.js
        Context: story
        Collapsed
        Icon: component
        Command: Open file [Ln 4, Col 15]

        - Primary
          Id: custom-title-prefix-second-level-csf-emptydefaultexport--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/EmptyDefaultExport.stories.js [Ln 8, Col 13]

      - EmptyTitle
        Id: custom-title-prefix-second-level-csf-emptytitle
        File: file:///mock/basedir/project/v6/src/autoTitle/csf/EmptyTitle.stories.js
        Context: story
        Collapsed
        Icon: component
        Command: Open file [Ln 4, Col 15]

        - Primary
          Id: custom-title-prefix-second-level-csf-emptytitle--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/EmptyTitle.stories.js [Ln 10, Col 13]

      - nested
        Id: custom-title-prefix-second-level-csf-nested
        File: file:///mock/basedir/project/v6/src/autoTitle/csf/nested/nested.stories.js
        Context: story
        Collapsed
        Icon: folder
        Command: Open file [Ln 4, Col 15]

        - WithoutTitle
          Id: custom-title-prefix-second-level-csf-nested-withouttitle
          File: file:///mock/basedir/project/v6/src/autoTitle/csf/nested/WithoutTitle.stories.js
          Context: story
          Collapsed
          Icon: component
          Command: Open file [Ln 4, Col 15]

          - Primary
            Id: custom-title-prefix-second-level-csf-nested-withouttitle--primary
            Context: story
            Icon: bookmark
            Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/nested/WithoutTitle.stories.js [Ln 10, Col 13]

        - Primary
          Id: custom-title-prefix-second-level-csf-nested--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/nested/nested.stories.js [Ln 10, Col 13]

      - withIndex
        Id: custom-title-prefix-second-level-csf-withindex
        File: file:///mock/basedir/project/v6/src/autoTitle/csf/withIndex/index.stories.js
        Context: story
        Collapsed
        Icon: component
        Command: Open file [Ln 4, Col 15]

        - Primary
          Id: custom-title-prefix-second-level-csf-withindex--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/withIndex/index.stories.js [Ln 10, Col 13]

    - Explicit title
      Id: custom-title-prefix-second-level-explicit-title
      File: file:///mock/basedir/project/v6/src/autoTitle/csf/ExtraTitle.stories.js
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 4, Col 15]

      - Primary
        Id: custom-title-prefix-second-level-explicit-title--primary
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/autoTitle/csf/ExtraTitle.stories.js [Ln 10, Col 13]

    - mdx
      Collapsed
      Icon: folder

      - MdxMeta
        Id: custom-title-prefix-second-level-mdx-mdxmeta--page
        File: file:///mock/basedir/project/v6/src/autoTitle/mdx/MdxMeta.stories.mdx
        Context: story
        Icon: document
        Command: Open file [Ln 2, Col 0]

      - MdxMetaAndStory
        Id: custom-title-prefix-second-level-mdx-mdxmetaandstory
        File: file:///mock/basedir/project/v6/src/autoTitle/mdx/MdxMetaAndStory.stories.mdx
        Context: story
        Collapsed
        Icon: component
        Command: Open file [Ln 2, Col 0]

        - Primary
          Id: custom-title-prefix-second-level-mdx-mdxmetaandstory--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/mdx/MdxMetaAndStory.stories.mdx [Ln 4, Col 0]

      - MdxStory
        Id: custom-title-prefix-second-level-mdx-mdxstory
        File: file:///mock/basedir/project/v6/src/autoTitle/mdx/MdxStory.stories.mdx
        Context: story
        Collapsed
        Icon: component
        Command: Open file

        - Primary
          Id: custom-title-prefix-second-level-mdx-mdxstory--primary
          Context: story
          Icon: bookmark
          Command: Open file file:///mock/basedir/project/v6/src/autoTitle/mdx/MdxStory.stories.mdx [Ln 2, Col 0]

- Example
  Expanded

  - MDX importing CSF
    Id: example-mdx-importing-csf
    File: file:///mock/basedir/project/v6/src/stories/CsfWithMdx.stories.mdx
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 4, Col 0]

    - Basic
      Id: example-mdx-importing-csf--basic
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfWithMdx.stories.mdx [Ln 10, Col 0]

    - Story name from MDX
      Id: example-mdx-importing-csf--basic-2
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfWithMdx.stories.mdx [Ln 20, Col 0]

    - Basic 3
      Id: example-mdx-importing-csf--basic-3
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfWithMdx.stories.mdx [Ln 24, Col 0]

  - Custom ID MDX
    Id: manually-specified-mdx-id
    File: file:///mock/basedir/project/v6/src/stories/CustomId.stories.mdx
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 2, Col 0]

    - Story
      Id: manually-specified-mdx-id--story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomId.stories.mdx [Ln 6, Col 0]

  - MDX Docs Only
    Id: example-mdx-docs-only--page
    File: file:///mock/basedir/project/v6/src/stories/DocsOnly.stories.mdx
    Context: story
    Icon: document
    Command: Open file [Ln 2, Col 0]

  - MDX with Stories
    Id: example-mdx-with-stories
    File: file:///mock/basedir/project/v6/src/stories/Mdx2WithStories.stories.mdx
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 11, Col 0]

    - MDX first story
      Id: example-mdx-with-stories--mdx-first-story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/Mdx2WithStories.stories.mdx [Ln 13, Col 0]

    - MDX second story
      Id: example-mdx-with-stories--mdx-second-story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/Mdx2WithStories.stories.mdx [Ln 17, Col 0]

    - MDX Story 3
      Id: example-mdx-with-stories--mdx-story-3
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/Mdx2WithStories.stories.mdx [Ln 21, Col 0]

  - Precedence
    Id: example-precedence
    File: file:///mock/basedir/project/v6/src/stories/Precedence.stories.mdx
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 3, Col 0]

    - Story with name, story, body
      Id: example-precedence--small
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/Precedence.stories.mdx [Ln 19, Col 0]

    - Story with name, body
      Id: example-precedence--story-with-name-body
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/Precedence.stories.mdx [Ln 25, Col 0]

  - Basic CSF
    Id: example-basic-csf
    File: file:///mock/basedir/project/v6/src/stories/BasicCsf.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 4, Col 15]

    - Primary
      Id: example-basic-csf--primary
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/BasicCsf.stories.js [Ln 14, Col 13]

    - Secondary
      Id: example-basic-csf--secondary
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/BasicCsf.stories.js [Ln 19, Col 13]

    - Large
      Id: example-basic-csf--large
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/BasicCsf.stories.js [Ln 24, Col 13]

    - Small
      Id: example-basic-csf--small
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/BasicCsf.stories.js [Ln 30, Col 13]

  - CSF v1
    Id: example-csf-v1
    File: file:///mock/basedir/project/v6/src/stories/CsfV1.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 0, Col 15]

    - Renamed via non-hoisted story
      Id: example-csf-v1--v-1-story-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfV1.stories.js [Ln 4, Col 13]

    - v2 name override
      Id: example-csf-v1--v-2-v-1-story-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfV1.stories.js [Ln 10, Col 13]

    - v1 name should be used
      Id: example-csf-v1--v-2-v-1-story-name-empty
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CsfV1.stories.js [Ln 17, Col 13]

  - Custom ID CSF
    Id: manually-specified-csf-story-id
    File: file:///mock/basedir/project/v6/src/stories/CustomId.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 0, Col 15]

    - Story
      Id: manually-specified-csf-story-id--story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomId.stories.js [Ln 5, Col 13]

  - Custom Name
    Id: example-custom-name
    File: file:///mock/basedir/project/v6/src/stories/CustomName.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 4, Col 15]

    - Custom story name
      Id: example-custom-name--custom-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 15, Col 13]

    - Custom story name override
      Id: example-custom-name--overridden
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 21, Col 13]

    - Custom story name var
      Id: example-custom-name--named-via-variable
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 28, Col 13]

    - Custom story name var in template string
      Id: example-custom-name--interpolated-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 35, Col 13]

    - Empty Name
      Id: example-custom-name--empty-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 38, Col 13]

    - Leading whitespace
      Id: example-custom-name--leading-whitespace
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/CustomName.stories.js [Ln 41, Col 13]

  - Deep
    Collapsed
    Icon: folder

    - Tree
      Collapsed
      Icon: folder

      - With
        Collapsed
        Icon: folder

        - Several
          Collapsed
          Icon: folder

          - Kinds
            Id: example-deep-tree-with-several-kinds
            File: file:///mock/basedir/project/v6/src/stories/DeepTree.stories.js
            Context: story
            Collapsed
            Icon: component
            Command: Open file [Ln 2, Col 15]

            - Story
              Id: example-deep-tree-with-several-kinds--story
              Context: story
              Icon: bookmark
              Command: Open file file:///mock/basedir/project/v6/src/stories/DeepTree.stories.js [Ln 9, Col 13]

  - Export Styles
    Id: example-export-styles
    File: file:///mock/basedir/project/v6/src/stories/ExportStyles.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 4, Col 15]

    - Normal Export
      Id: example-export-styles--normal-export
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/ExportStyles.stories.js [Ln 14, Col 6]

    - Alternative Export Name
      Id: example-export-styles--alternative-export-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/ExportStyles.stories.js [Ln 19, Col 6]

    - Override name 3
      Id: example-export-styles--export-name-overridden
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/ExportStyles.stories.js [Ln 24, Col 6]

  - Form
    Id: example-form--form
    File: file:///mock/basedir/project/v6/src/stories/Form.stories.js
    Context: story
    Icon: bookmark
    Command: Open file [Ln 6, Col 13]

  - Hoisting
    Id: example-hoisting--hoisting
    File: file:///mock/basedir/project/v6/src/stories/Hoisting.stories.js
    Context: story
    Icon: bookmark
    Command: Open file [Ln 2, Col 13]

  - JSX extension
    Id: example-jsx-extension
    File: file:///mock/basedir/project/v6/src/stories/JsxExtension.stories.jsx
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 0, Col 15]

    - Story
      Id: example-jsx-extension--story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/JsxExtension.stories.jsx [Ln 4, Col 13]

  - Merge
    Id: file:///mock/basedir/project/v6/src/stories/MergeOne.stories.js
    File: file:///mock/basedir/project/v6/src/stories/MergeOne.stories.js
    Collapsed
    Icon: component
    Command: Open file

    - One
      Id: example-merge--one
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/MergeOne.stories.js [Ln 4, Col 13]

    - Two
      Id: example-merge--two
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/MergeTwo.stories.js [Ln 4, Col 13]

  - With Same Name
    Id: example-with-same-name
    File: file:///mock/basedir/project/v6/src/stories/SameName.stories.js
    Context: story
    Collapsed
    Icon: folder
    Command: Open file [Ln 4, Col 15]

    - Same Name
      Id: example-with-same-name--same-name
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/SameName.stories.js [Ln 11, Col 13]

    - Same Name
      Id: example-with-same-name-same-name
      File: file:///mock/basedir/project/v6/src/stories/SameNameChild.stories.js
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 4, Col 15]

      - Nested Story
        Id: example-with-same-name-same-name--nested-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/SameNameChild.stories.js [Ln 11, Col 13]

  - Title Interpolation
    Id: example-title-interpolation
    File: file:///mock/basedir/project/v6/src/stories/TitleInterpolation.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 6, Col 15]

    - Primary
      Id: example-title-interpolation--primary
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/TitleInterpolation.stories.js [Ln 13, Col 13]

- Test
  Expanded

  - Exclusions
    Collapsed
    Icon: folder

    - Excluded
      Id: test-exclusions-excluded
      File: file:///mock/basedir/project/v6/src/stories/Excluded.stories.js
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 3, Col 15]

      - Other Export
        Id: test-exclusions-excluded--other-export
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/Excluded.stories.js [Ln 1, Col 13]

    - Included
      Id: test-exclusions-included
      File: file:///mock/basedir/project/v6/src/stories/Included.stories.js
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 3, Col 15]

      - Included Export
        Id: test-exclusions-included--included-export
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/Included.stories.js [Ln 0, Col 13]

  - Story Declarations
    Id: test-story-declarations
    File: file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js
    Context: story
    Collapsed
    Icon: component
    Command: Open file [Ln 0, Col 6]

    - Const Declarator
      Id: test-story-declarations--const-declarator
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 6, Col 13]

    - Separate Declarator
      Id: test-story-declarations--separate-declarator
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 8, Col 6]

    - Separate declarator with story name
      Id: test-story-declarations--separate-declarator-2
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 9, Col 6]

    - Multiple One
      Id: test-story-declarations--multiple-one
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 14, Col 13]

    - Multiple two (given name)
      Id: test-story-declarations--multiple-two
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 15, Col 2]

    - Renamed Function Story
      Id: test-story-declarations--function-story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 19, Col 16]

    - Separate Function Story
      Id: test-story-declarations--separate-function-story
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 24, Col 9]

    - Separate function story with story name
      Id: test-story-declarations--separate-function-story-2
      Context: story
      Icon: bookmark
      Command: Open file file:///mock/basedir/project/v6/src/stories/StoryDeclarations.stories.js [Ln 28, Col 9]

  - TypeScript
    Collapsed
    Icon: folder

    - Sample
      Id: test-typescript-sample
      File: file:///mock/basedir/project/v6/src/stories/TypeScript.stories.tsx
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 16, Col 15]

      - Type Script Story
        Id: test-typescript-sample--type-script-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScript.stories.tsx [Ln 2, Col 13]

      - Type Script Typed Story
        Id: test-typescript-sample--type-script-typed-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScript.stories.tsx [Ln 4, Col 13]

      - Type Script Typed Generic Story
        Id: test-typescript-sample--type-script-typed-generic-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScript.stories.tsx [Ln 8, Col 13]

      - Type Script Type Assertion Story
        Id: test-typescript-sample--type-script-type-assertion-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScript.stories.tsx [Ln 12, Col 13]

    - Default as Meta
      Id: test-typescript-default-as-meta
      File: file:///mock/basedir/project/v6/src/stories/TypeScriptAsMeta.stories.ts
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 4, Col 15]

      - Type Script Story
        Id: test-typescript-default-as-meta--type-script-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScriptAsMeta.stories.ts [Ln 2, Col 13]

    - Typed Identifier
      Id: test-typescript-typed-identifier
      File: file:///mock/basedir/project/v6/src/stories/TypeScriptIdentifier.stories.ts
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 4, Col 6]

      - Type Script Story
        Id: test-typescript-typed-identifier--type-script-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScriptIdentifier.stories.ts [Ln 2, Col 13]

    - Identifier Default Exported as Meta
      Id: test-typescript-identifier-default-exported-as-meta
      File: file:///mock/basedir/project/v6/src/stories/TypeScriptIdentifierDefaultAsMeta.stories.ts
      Context: story
      Collapsed
      Icon: component
      Command: Open file [Ln 5, Col 6]

      - Type Script Story
        Id: test-typescript-identifier-default-exported-as-meta--type-script-story
        Context: story
        Icon: bookmark
        Command: Open file file:///mock/basedir/project/v6/src/stories/TypeScriptIdentifierDefaultAsMeta.stories.ts [Ln 2, Col 13]
`;
